### PR TITLE
add multi stream allocations benchmark.

### PR DESCRIPTION
**Pull request by cwharris**
This PR introduces a new benchmark to measure the effect of the pool allocator in situations which require concurrent kernel execution. The benchmark reveals that the pool allocator prevents concurrent kernel execution in multiple non-default streams unless the pool allocator already has enough memory reserved for each stream prior to any allocation attempts.

legend: `BM_MultiStreamAllocations/<threads in pool>/<kernels to launch>/<prewarm>`
```
-------------------------------------------------------------------------------------------
Benchmark                                 Time             CPU   Iterations UserCounters...
-------------------------------------------------------------------------------------------
BM_MultiStreamAllocations/cuda/4/4/0             2568 us         2564 us          270 items_per_second=1.55987k/s
BM_MultiStreamAllocations/cuda/4/4/1             2568 us         2562 us          272 items_per_second=1.5612k/s
BM_MultiStreamAllocations/cuda_async/4/4/0        536 us          536 us         1301 items_per_second=7.45966k/s
BM_MultiStreamAllocations/cuda_async/4/4/1        537 us          537 us         1303 items_per_second=7.45012k/s
BM_MultiStreamAllocations/pool_mr/4/4/0          2094 us         2094 us          334 items_per_second=1.90985k/s
BM_MultiStreamAllocations/pool_mr/4/4/1           533 us          533 us         1310 items_per_second=7.49885k/s
BM_MultiStreamAllocations/arena/4/4/0            2105 us         2105 us          332 items_per_second=1.90017k/s
BM_MultiStreamAllocations/arena/4/4/1            2113 us         2113 us          332 items_per_second=1.89299k/s
BM_MultiStreamAllocations/binning/4/4/0          2104 us         2104 us          333 items_per_second=1.9009k/s
BM_MultiStreamAllocations/binning/4/4/1           534 us          534 us         1307 items_per_second=7.49373k/s
```

The benchmark works by running a compute-bound kernel in a very small launch configuration (1 block 1 thread) multiple times against a stream pool. As the size of the stream pool increases, one should expect the overall performance to improve. This is only the case when the memory pool has been prewarmed for each stream - otherwise, sync-and-steal behavior comes in to play, forcing any already-launched work to complete prior to allocating the next buffer, therefore preventing any kernel overlap.

undesirable synchronization:
-
cold start

desirable kernel overlap:
-
warm start

### CODE REVIEW

File: benchmarks/multi_stream_allocations/multi_stream_allocations_bench.cu [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    }
+  }
+
+  *out = static_cast<int64_t>(clock_current);
+}
+
+static void BM_MultiStreamAllocations(benchmark::State& state)
+{
+  auto cuda_mr = rmm::mr::cuda_memory_resource{};
+  auto mr      = rmm::mr::pool_memory_resource<rmm::mr::cuda_memory_resource>{&cuda_mr};
```
Comment (harrism) [inline benchmarks/multi_stream_allocations/multi_stream_allocations_bench.cu:44]: I think  that you should make the MR a parameter, not hard coded. See the `random_allocations` benchmark for a way to do this. I wouldn't put in all the machinery to benchmark all MRs like that one has, just start with a pool<cuda> MR for now.

File: benchmarks/multi_stream_allocations/multi_stream_allocations_bench.cu [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    }
+
+    cudaDeviceSynchronize();
+  }
+
+  state.SetItemsProcessed(state.iterations() * num_kernels);
+
+  rmm::mr::set_current_device_resource(nullptr);
+}
+BENCHMARK(BM_MultiStreamAllocations)
```
Comment (harrism) [inline benchmarks/multi_stream_allocations/multi_stream_allocations_bench.cu:76]: Look at `random_allocations.cpp` for how it has a command line option to run the function raw without gbenchmark to make profiling easier (gbench runs many times, which is superfluous and confusing when searching for the relevant portion of a profile).

**Review by harrism — CHANGES_REQUESTED** [review]
Thanks for creating this. Looks good. Just some suggestions on how to make it easier to use for profiling and easier to extend to other MRs in the future.

**Review by harrism — APPROVED** [review]
Love it.

File: benchmarks/CMakeLists.txt [hunk 3]
```diff
@@ -79,3 +79,10 @@ set(CUDA_STREAM_POOL_BENCH_SRC
     "${CMAKE_CURRENT_SOURCE_DIR}/cuda_stream_pool/cuda_stream_pool_bench.cpp")
 
 ConfigureBench(CUDA_STREAM_POOL_BENCH "${CUDA_STREAM_POOL_BENCH_SRC}")
+
+# multi stream allocations
```
Comment (robertmaynard) [inline benchmarks/CMakeLists.txt:83]: With https://github.com/rapidsai/rmm/pull/842 merged, you should update the benchmark/CMakeLists.txt to match the new style.

**Review by robertmaynard — APPROVED** [review]

File: benchmarks/multi_stream_allocations/multi_stream_allocations_bench.cu [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+  if (name == "binning") {
+    BENCHMARK_CAPTURE(BM_MultiStreamAllocations, binning, &make_binning)  //
+      ->Apply(benchmark_range);
+    return;
+  }
+
+  std::cout << "Error: invalid memory_resource name: " << name << std::endl;
+}
+
+void run_profile(std::string resource_name, int kernel_count, int stream_count, bool prewarm)
```
Comment (jrhemstad) [inline benchmarks/multi_stream_allocations/multi_stream_allocations_bench.cu:177]: Is this needed? Can't you do the same thing via GBench command line args?

Comment (harrism) [reply benchmarks/multi_stream_allocations/multi_stream_allocations_bench.cu:177]: Not sure what you mean by "this". gbench runs multiple times no matter what. We want a way to run only once for profiling.

**Comment by harrism** [comment]
@cwharris failing cmake style

File: benchmarks/multi_stream_allocations/multi_stream_allocations_bench.cu (hunk 4 above)
Comment (jrhemstad) [reply benchmarks/multi_stream_allocations/multi_stream_allocations_bench.cu:177]: > We want a way to run only once for profiling.

Right, that's what I meant. I thought there was a "num_iterations` gbench option. 

Related https://github.com/NVIDIA/nvbench/issues/10

Comment (cwharris) [reply benchmarks/multi_stream_allocations/multi_stream_allocations_bench.cu:177]: I haven't found an option in gbench to limit the number of iterations. I looked briefly prior to implementing it this way, but gbench documentation is... not great?

Comment (harrism) [reply benchmarks/multi_stream_allocations/multi_stream_allocations_bench.cu:177]: gbenchmark does not have a way to control number of iterations. Here's one of the authors explanations of why. https://stackoverflow.com/a/61888885/749748

**Comment by harrism** [comment]
@gpucibot merge

**Pull request merged**